### ISSUE TITLE
with clang, [[nodiscard]] is emitted also for -std=c++11

### ISSUE THREAD

**Issue by botovq** [protocolbuffers/protobuf#9181]
**What version of protobuf and what language are you using?**
Version: v3.19.1
Language: C++

**What operating system (Linux, Windows, ...) and version?**

OpenBSD -current

**What runtime / compiler are you using (e.g., python version or gcc version)**

clang 11.1.0

**What did you do?**

Build protobuf 3.19.1, then try to build protozero 1.7.0.

**What did you expect to see**

protozero 1.7.0 builds

**What did you see instead?**

The build failed because protozero builds with `-pedantic -std=c++11 -Werror`. On the one hand, this is enough to have `PROTOBUF_NODISCARD` defined as `[[nodiscard]]`. On the other hand, this then trips over `-Werror` due to `-Wc++17-extensions`:

```
[79/90] /tmp/pobj/protozero-1.7.0/bin/c++  -I/tmp/pobj/protozero-1.7.0/protozero-1.7.0/include -I/tmp/pobj/protozero-1.7.0/protozero-1.7.0/test/include -isystem /tmp/pobj/protozero-1.7.0/protozero-1.7.0/test/catch -isystem /usr/local/include -isystem test -O2 -pipe  -std=c++17 -DNDEBUG   -std=c++11 -Wall -Wextra -pedantic -Wsign-compare -Wunused-parameter -Wno-float-equal -Wno-covered-switch-default -Werror -pthread -MD -MT test/CMakeFiles/writer_tests.dir/t/bytes/bytes_testcase.pb.cc.o -MF test/CMakeFiles/writer_tests.dir/t/bytes/bytes_testcase.pb.cc.o.d -o test/CMakeFiles/writer_tests.dir/t/bytes/bytes_testcase.pb.cc.o -c test/t/bytes/bytes_testcase.pb.cc
FAILED: test/CMakeFiles/writer_tests.dir/t/bytes/bytes_testcase.pb.cc.o /tmp/pobj/protozero-1.7.0/bin/c++  -I/tmp/pobj/protozero-1.7.0/protozero-1.7.0/include -I/tmp/pobj/protozero-1.7.0/protozero-1.7.0/test/include -isystem /tmp/pobj/protozero-1.7.0/protozero-1.7.0/test/catch -isystem /usr/local/include -isystem test -O2 -pipe  -std=c++17 -DNDEBUG   -std=c++11 -Wall -Wextra -pedantic -Wsign-compare -Wunused-parameter -Wno-float-equal -Wno-covered-switch-default -Werror -pthread -MD -MT test/CMakeFiles/writer_tests.dir/t/bytes/bytes_testcase.pb.cc.o -MF test/CMakeFiles/writer_tests.dir/t/bytes/bytes_testcase.pb.cc.o.d -o test/CMakeFiles/writer_tests.dir/t/bytes/bytes_testcase.pb.cc.o -c test/t/bytes/bytes_testcase.pb.cc
In file included from test/t/bytes/bytes_testcase.pb.cc:4:
test/t/bytes/bytes_testcase.pb.h:190:3: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
  PROTOBUF_NODISCARD std::string* release_s();
  ^
/usr/local/include/google/protobuf/port_def.inc:463:30: note: expanded from macro 'PROTOBUF_NODISCARD'
#define PROTOBUF_NODISCARD [[nodiscard]]
                             ^
1 error generated.
ninja: build stopped: subcommand failed.
```

**Anything else we should know about your project / environment**

It seems `#if __has_cpp_attribute(nodiscard)` is insufficient to determine whether `[[nodiscard]]` is actually allowed. A similar issue was also reported here: https://github.com/nlohmann/json/issues/1535

This could be avoided if PROTOBUF_NODISCARD were only defined to `[[nodiscard]]` with

```c++
#if __has_cpp_attribute(nodiscard) && PROTOBUF_CPLUSPLUS_MIN(201703L)
#define PROTOBUF_NODISCARD [[nodiscard]]
#elif
...
```

or similar.

**Comment by thomas-olszamowski** [comment]
Confirming same problem with: 
Apple clang version 13.0.0
and protobuf 3.19.2

**Comment by ezzieyguywuf** [comment]
To add some more context: it appears that 624d29d83306f3ce2c7e092dd44a891e04215e67 added [these lines](https://github.com/protocolbuffers/protobuf/blame/master/src/google/protobuf/port_def.inc#L461-L469) to `port_def.inc`, which is where the error is originating from.

I think @botovq 's suggestion to extend the logic in the preprocessor check should help to mitigate this.

**Comment by ezzieyguywuf** [comment]
Here's how you can reproduce the error with a minimal example:

```sh
protoc --version # must be >= 3.18.1
```

```txt
# test.proto
syntax = "proto2";

package Nothing;

message Message {
  optional string something = 1;
}
```
```cpp
// main.cpp
#include "test.pb.h"

int main(){}
```

```sh
# first, generate the protobuf code
protoc -I=. --cpp_out=. test.proto

# this works
clang++ -std=c++17 -pedantic -Werror main.cpp

# this does not
clang++ -std=c++14 -pedantic -Werror main.cpp
```

**Comment by ezzieyguywuf** [comment]
This bug appears to be resolved by ab4585a6956675ce14a1cba5d321fde980bbf12b

See [this line](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/port_def.inc#L476)

**Comment by oxoocoffee** [comment]
Confirming the same problem with:
Apple clang version 13.0.0 (clang-1300.0.29.30)
protoc --version
libprotoc 3.19.4

Building code with c++11

**Comment by ezzieyguywuf** [comment]
> Confirming the same problem with:
> Apple clang version 13.0.0 (clang-1300.0.29.30)
> protoc --version
> libprotoc 3.19.4

I think this should be resolved in the next release.

You could try compiling protoc from the latest source to confirm

**Comment by botovq** [comment]
As pointed out by @ezzieyguywuf this issue is resolved in protobuf 3.20 and later.
